workers/unity/Packages/com.improbable.gdk.testutils/TestLogDispatcher.cs

```diff
@@ -0,0 +1,148 @@
+        }
+
+        public void HandleLog(LogType type, LogEvent logEvent)
+        {
+            if (currentExpectingScope != null)
+            {
+                currentExpectingScope.CheckIncomingLog(type, logEvent);
+                return;
+            }
+
+            if (type == LogType.Error || type == LogType.Exception)
+            {
+                Assert.Fail($"Encountered error log outside of an expecting scope: [{type}] - {logEvent}");
+            }
+        }
+
+        /// <summary>
+        ///     Creates and returns an disposable ExpectingScope object. This is intended to be used with a using block.
+        /// </summary>
+        /// <returns>An ExpectingScope instance.</returns>
+        /// <exception cref="InvalidOperationException">
+        ///     Throws if you already have an un-disposed ExpectingScope from this logger.
+        /// </exception>
+        public ExpectingScope EnterExpectingScope()
+        {
+            if (currentExpectingScope != null)
+            {
+                throw new InvalidOperationException("Cannot ");
+            }
+
+            currentExpectingScope = new ExpectingScope(this);
+            return currentExpectingScope;
+        }
+
+        public void Dispose()
+        {
+            if (currentExpectingScope != null)
+            {
+                throw new InvalidOperationException(
+                    "Cannot Dispose a TestLogDispatcher while there is an outstanding ExpectingScope");
+            }
+        }
+
+        public class ExpectingScope : IDisposable
+        {
+            private readonly Queue<LogStructure> expectedLogs = new Queue<LogStructure>();
+            private readonly Queue<(LogType, LogEvent)> unexpectedLogs = new Queue<(LogType, LogEvent)>();
+
+            private TestLogDispatcher dispatcher;
+
+            internal ExpectingScope(TestLogDispatcher dispatcher)
+            {
+                this.dispatcher = dispatcher;
+            }
+
+            public void Expect(LogType type, params string[] dataKeys)
+            {
+                expectedLogs.Enqueue(new LogStructure(type, dataKeys));
+            }
+
+            public void Dispose()
+            {
+                dispatcher.currentExpectingScope = null;
+
+                Assert.AreEqual(0, expectedLogs.Count);
+
+                var unexpectedLogsString = string.Join("\n", unexpectedLogs.Select(log =>
+                {
+                    var (logType, logEvent) = log;
+                    return $"[{logType}] - {logEvent}";
+                }));
+                Assert.AreEqual(0, unexpectedLogs.Count,
+                    $"Received unexpected errors or exceptions : {unexpectedLogsString}");
+            }
+
+            internal void CheckIncomingLog(LogType type, LogEvent logEvent)
+            {
+                if (expectedLogs.Count > 0)
+                {
+                    if (expectedLogs.Peek().DoesMatchLog(type, logEvent))
```

Comment:
can merge the two ifs into one

Reply:
I was worried it would make the if statement too long - but sure :)